share/wake/lib/rust_wake/cargo.wake

```diff
@@ -25,21 +29,193 @@ export target findCargoPath Unit =
+        False -> None
+
+    # Locates all the sources needed to build a CargoCrate
+    def findSources (crate: CargoCrate): Result (List Path) Error =
+        def root =
+            crate.getCargoCrateRoot.getPathName
+            | simplify
+
+        require Pass toml = source "{root}/Cargo.toml"
+        require Pass lock = source "{root}/Cargo.lock"
+        require Pass srcs = sources "{root}/src" `.*`
+
+        Pass (toml, lock, srcs)
+
+    def (CargoToolchain cargo config rootCrate localCrates locked flags) = toolchain
+    def (CargoCrate rootPath binaryName) = rootCrate
+
+    def root =
+        rootPath.getPathName
+        | simplify
+
+    # wake-format off
+    def cmd =
+        cargo,
+        subcommand,
+        "--quiet",
+        "--bin", binaryName,
+        ((buildConfigToFlag config) | optionToList) ++
+        ((lockedToFlag locked) | optionToList) ++
+        # To not incur a O(n^2) cost, flags are appended in reverse of the callers expectations.
+        # Reverse here to correct them.
+        (flags | reverse)
+
+    require Pass sources =
+        (rootCrate, localCrates)
+        | map findSources
+        | findFail
+        | rmap flatten
+
+    def job =
+        makeExecPlan cmd (sources)
+        | setPlanLabel "cargo {subcommand}: {binaryName} [{buildConfigToString config}]"
+        | setPlanDirectory root
+        | setPlanFilterOutputs (matches fnOutputsFilter)
+        | setPlanIsAtty True
+        # Adds cargo and rustc to the environment
+        | prependPlanPath (dirname cargo)
         | runJobWith defaultRunner
+
+    require True = job.isJobOk
+    else failWithError "Failed to build with cargo"
+
+    require Pass outputs =
+        job
         | getJobOutputs
 
-    Pass outputs
+    require (bin, Nil) = outputs
+    else failWithError "More than one output returned when building with cargo"
+
+    Pass bin
+
+# Builds a cargo project from the CargoToolchain
+# 
+# Cargo will not reuse work because the sandbox will not contain its extra info, inside of target.
+# TODO: Add an environment variable to opt-in to localRunner cargo builds if we detect
+#       existing outputs in {root}/target/release
+export def cargoBuild (toolchain: CargoToolchain): Result Path Error =
+    def (CargoToolchain _ config rootCrate _ _ _) = toolchain
+    def (CargoCrate rootPath binaryName) = rootCrate
+
+    def root =
+        rootPath.getPathName
+        | simplify
+
+    require Pass regex = stringToRegExp "{root}/target/{buildConfigToString config}/{binaryName}"
+
+    toolchain
+    | cargoBuildCommon "build" regex
+
+# Builds a cargo test binary from the CargoToolchain
+# 
+# Cargo will not reuse work because the sandbox will not contain its extra info, inside of target.
+# TODO: Add an environment variable to opt-in to localRunner cargo builds if we detect
+#       existing outputs in {root}/target/release
+export def cargoTestBuild (toolchain: CargoToolchain): Result Path Error =
```

Comment:
I'm torn, I also really want `cargoTest` but I think it would confused rust developers because typing `cargo test` in a shell will build and run the test while this def only builds the test.

You could easily convince me to use `cargoTest` though

Reply:
updated the data structure a bit and now there is just one entry function `cargoBuildWith`